### DISCUSSION TITLE
VEML60, how  to change unit

### DISCUSSION THREAD

**Discussion by ThomasT500** [arendst/Tasmota#15138] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
VEML60 reporting wrong UV Power value 

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x ] Device used (e.g., Sonoff Basic): _____
- [x ] Tasmota binary firmware version number used: _____
  - [x ] Pre-compiled
  - [x] Self-compiled
- [x ] Flashing tools used: Tasmotizer
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
ESP32 with Tasmota:
Program Version	11.0.0.2(allsensors)
Build Date & Time	2022-02-21T22:07:30


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
 Rule1: on tele-VEML6070#UvIndexPower do publish domoticz/in {"idx":61,"nvalue":0,"svalue":"%value%"} endon


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
None

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
None

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Tried new flash and new configuration, same result

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
Publish this value in the right dataformat ór - better - publish it inW/m2 to get a more readable output for sending to e.g. Domoticz
and in this way also is published e.g. by “Buitenradar” and others

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
![TasmotaMainMenuwithVEML6070](https://user-images.githubusercontent.com/78922044/158155544-fe5b61d1-2317-467b-9ff3-8dad8249e450.PNG)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._
Tried by myself with NO succes:  
Rule1: on tele-VEML6070#UvIndexPower do publish domoticz/in {"idx":61,"nvalue":0,"svalue":"%value%"*1000} endon

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by barbudor** [comment]
You don't explain why the value is incorrect
The UV Index is 0.98 which is totally coherent with a UV Power of 0.025 W/m2

**Reply by ThomasT500** [reply]
You are right but instead of W/m2 it has to be W/**Cm**2

**Reply by barbudor** [reply]
> You are right but instead of W/m2 it has to be W/**Cm**2

Why ?
It has been like that for 4 years, no-one complained.
Changing unit would be a major breaking change for existing users.

**Comment by sfromis** [comment, marked as answer]
> Rule1: on tele-VEML6070#UvIndexPower do publish domoticz/in {"idx":61,"nvalue":0,"svalue":"%value%"*1000} endon

You can't have expression mixed with the text of a command, but instead first calculate the value in one rule, and then refer it in another.

And the unit of watt per square meter is totally consistent with the norm for measuring solar irradiance. From Wikipedia:
> The solar irradiance is measured in watt per square metre (W/m2) in SI units.

Generally: When raising an issue **do not** just ignore most of the templated questions you're asked to fill in. They are there for a purpose.

**Reply by ThomasT500** [reply]
This was my first post here and I will respect the  the templated questions in future.

You are right, but the Tasmota value is reporting the **W/Cm2** I think with the unit in W/m2.
My test setup is the VEML6070 in a borosilicate laboratory tube (is uv transparant)

=> My experiments with the "Tasmota Rules Arithmetic commands" <  MULT1 VAR >  where NOT succesfull up to now ;-(
Can you help me with an example please?

**Reply by sfromis** [reply]
`Mult1 1000` would multiply `var1` by 1000, but if the value is indeed `W/cm²`, the multiplier to `W/m²` would be 10000, not 1000. However, if you're right, the driver should be fixed instead of using rules to postprocess.

Variables do not have names, but numbers, and `VAR` is again not a name, but a keyword which in some contexts indicate what variable number you want.

Full sun on a clear day should be up to 1000 W/m², which thus would be 0.1 W/cm², meaning that the displayed value of 0.025 would be approx ¼ of full sun, if it is indeed W/cm² instead of W/m². Is that in line with what you're doing? Of course full irradiance is more than just UV, but the energic UV part of the spectrum should be a sizable part of the irradiance.

**Reply by barbudor** [reply]
What I read on Wikipedia is
> The UV power spectrum (expressed as watts per square meter per nanometer of wavelength) is therefore multiplied by a [weighting curve](https://en.wikipedia.org/wiki/Weighting_curve) known as the erythemal action spectrum, and the result integrated over the whole spectrum. This gives a weighted figure (sometimes called Diffey-weighted UV irradiance, or DUV, or erythemal dose rate) typically around 250 mW/m2 in midday summer-sunlight. For convenience this is divided by 25 mW/m2 to produce an index[[8]](https://en.wikipedia.org/wiki/Ultraviolet_index#cite_note-uvicalc-8)[[9]](https://en.wikipedia.org/wiki/Ultraviolet_index#cite_note-UVIndex_calculation-9) nominally from 0 to 11+, though ozone depletion is now resulting in higher values.

So a summer sunny day is 0.25 W/m2 not 1000 W/m2
As the UV index is made by dividing the power by 25mW, an index of 1 means 0.025 W/m2

**Reply by sfromis** [reply]
As the index is up to about 11, a clear summer day should be up to 11 times 0.25, about 2.75 W/m² of UV. As I explained, 1000 W/m² is all irradiation, not just UV.

Anyway, this is not consistent with the expectation of having to multiply the UV measurement by 10000. Hence my unanswered question about the numbers being in line with what Thomas is doing. Better have specifics instead of assertions about "wrong" without any avenue for getting to a common understanding.

**Reply by barbudor** [reply]
@sfromis, I think you get it wrong by a factor of 10
Index 1 is 25 mW/m2 or 0.025 W/m2
Index 11 is 275 mW/m2 or 0.275 W/m2

**Reply by sfromis** [reply]
Yeah, we both got  it wrong by a factor of 10 :-)

**Reply by ThomasT500** [reply]
Thank you very much for all this background information Barbudor.
I have no doubt that this is all right.
But I have only a "problem" with the "UV Power" display ;-)

**Comment by ThomasT500** [comment]
m2 is indeed 100x100=10.000cm2
I am afraid in this case the driver need to be fixed as in other Tasmota Google information you find Tasmota main screen prints with uW/cm2.

For now I only want post processing the value. I can “distillate” the Var1 but can not "Mult1 100.000" this value and send it to Domoticz.

1. Can we change the mistake I made in the Topic name and change this in VEML6070, so more people can find our solution

1. Can you help me with an example formula ffor post processing the value and send this to Domoticz

Thnx. Thomas

**Reply by sfromis** [reply]
Please be very specific about why you see the number given as "wrong", how you get to that conclusion, and what values you'd expect on a clear summer day.

**Comment by ThomasT500** [comment]
=> For yet more information about this subject: See also discussion: https://github.com/arendst/Tasmota/issues/3844

My screen:
![1MainMenuWm2](https://user-images.githubusercontent.com/78922044/158359341-dab941ce-991d-4ad2-9c46-42d0d44630e4.PNG)

Screen from other github 3844 discussion:
![2MainScreenMicoWattM2](https://user-images.githubusercontent.com/78922044/158359401-13803f24-0626-400d-b9ea-be21c388b0f5.png)

Data screen from offical Dutch weatherstation "Buienradar":
![3Buienradar](https://user-images.githubusercontent.com/78922044/158359412-dfa4d595-65d0-4e36-9267-7b5a14184e85.png)

Although W/cm2 scientifically is right most people with sun panels on their roof talk about W/m2. This value also appeals more to the imagination of people.
The Tasmota “UV Power” figures now don’t show up really nice in Domoticz.

What I want:
1.	Correct the Tasmota “UV Power” in W/CM2
2.	Is it an idea to compile a choice for showing up the data in W/cm2 ór W/m2

3.	**For now: can anybody help me with a formula for post processing the “UV Power” value and multiply it with 10.000**
_What I have done myself for this:_
I can define Var1
When in console I give as input Mult1 10.000 this gives a correct result, but I don’t know how to “transport/send” this result to Domoticz (or any other MQTT client).
(I am not an programming specialist but an enthusiastic electronics hobbyist)


Thnx. Thomas

**Reply by arendst** [reply]
To send any data to Domoticz simply use a command like ``dzsend1 418,%var1%`` where ``418`` is the domoticz sensor index.

**Reply by sfromis** [reply]
Are you now saying that there is nothing wrong about what Tasmota does, you just prefer a different unit instead of W/m²?

**Reply by ThomasT500** [reply]
Thank you very much Theo for this answer from "the Master Himself" I feel myself very honoured!
I tried this command in the lunch break with - partly – success!
May I kindly ask you for a latest small “push in the back “.
=> How can I Var1 Mult1 by 10.000
Thnx. and enjoy your meal
(my example is in BME, but I can "translate this by myself to UV Power of course)

Rule2 on tele-BME280#Temperature do publish domoticz/in {"idx":51,"nvalue":0,"svalue":"%value%"} endon
on tele-BME280#Humidity do publish domoticz/in {"idx":52,"nvalue":0,"svalue":"%value%"} endon
on BME280#Temperature do Var1 %value% endon
on BME280#Temperature do dzsend 63,%var1% endon

**Reply by ThomasT500** [reply]
Sfromis Tasmota is a Github project with Theo as developer. We all can help him to improve Tasmota. I am not a programmer otherwise I perhaps could solve this ( "small") problem. Also proffesional software has almost always small bugs. My "problem" is (almost) solved for now, I am shure Theo or any otherone will help to improve this "bug".

**Reply by ThomasT500** [reply]
Dear Theo and other fellowmates, 
Hurray, I found the solution!
Below how you can multply a sensor value by any "factor" you want.
I found out that the order you fill in the Rule lines in this case is important.

Rule2 on tele-BME280#Temperature do publish domoticz/in {"idx":51,"nvalue":0,"svalue":"%value%"} endon
on tele-BME280#Humidity do publish domoticz/in {"idx":52,"nvalue":0,"svalue":"%value%"} endon
on BME280#Temperature do Var1 %value% endon
on BME280#Temperature do Mult1 100 endon
on BME280#Temperature do dzsend 63,%Var1% endon

Theo, again thank you very much for your input

And of course all other ones too ;-)

**Reply by ThomasT500** [reply]
See and convince yourself ;-)
Theo, 
The data upload from the _**calculated**_ UV Power _**from the Rule**_ to Domoticz is very "nervous" now. 
Is there any way to _reduce this "upload rate"_ to Domoticz?
![DomoticzScreen](https://user-images.githubusercontent.com/78922044/158379957-5c4b3f5e-8766-41d5-bd37-c357c194c3d2.PNG)

**Reply by arendst** [reply]
Yes. Use teleperiod data instead of immediate data in your rules.
